File: python/open_guardrail/guards/output_truncation.py
"""Detect if output was likely truncated or incomplete."""

import re
import time

from open_guardrail.core import GuardResult


def _ends_mid_sentence(text: str) -> bool:
    trimmed = text.rstrip()
    if not trimmed:
        return False
    return bool(re.match(r"[a-zA-Z,;:\-]", trimmed[-1]))


def _has_unclosed_brackets(text: str) -> bool:
    opens = {"(": ")", "[": "]", "{": "}"}
    stack: list[str] = []
    for ch in text:
        if ch in opens:
            stack.append(opens[ch])
        elif ch in ")]}":
            if stack and stack[-1] == ch:
                stack.pop()
    return len(stack) > 0


def _has_unclosed_quotes(text: str) -> bool:
    return text.count("'") % 2 != 0 or text.count('"') % 2 != 0


def _has_incomplete_code_block(text: str) -> bool:
    return text.count("```") % 2 != 0


def _ends_with_ellipsis(text: str) -> bool:
    trimmed = text.rstrip()
    return trimmed.endswith("...") or trimmed.endswith("\u2026")


def _has_incomplete_list(text: str) -> bool:
    lines = text.rstrip().split("\n")
    last = lines[-1].strip() if lines else ""
    return bool(
        re.match(r"^\d+\.\s*$", last)
        or re.match(r"^[-*]\s*$", last)
    )


class _OutputTruncation:
    def __init__(self, *, action: str = "warn") -> None:
        self.name = "output-truncation"
        self.action = action

    def check(self, text: str, stage: str = "output") -> GuardResult:
        start = time.perf_counter()
        signals: list[str] = []

        if _ends_mid_sentence(text):
            signals.append("ends-mid-sentence")
        if _has_unclosed_brackets(text):
            signals.append("unclosed-brackets")
        if _has_unclosed_quotes(text):
            signals.append("unclosed-quotes")
        if _has_incomplete_code_block(text):
            signals.append("incomplete-code-block")
        if _ends_with_ellipsis(text):
            signals.append("ends-with-ellipsis")
        if _has_incomplete_list(text):
            signals.append("incomplete-list")

        triggered = len(signals) > 0
        score = min(len(signals) / 3, 1.0) if triggered else 0.0
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="output-truncation",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Output appears truncated"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details={"signals": signals} if triggered else None,
        )


def output_truncation(
    *, action: str = "warn"
) -> _OutputTruncation:
    return _OutputTruncation(action=action)

File: packages/guards/src/output-truncation.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface OutputTruncationOptions {
  action: 'block' | 'warn';
}

function endsMidSentence(text: string): boolean {
  const trimmed = text.trimEnd();
  if (trimmed.length === 0) return false;
  const last = trimmed[trimmed.length - 1];
  return /[a-zA-Z,;:\-]/.test(last);
}

function hasUnclosedBrackets(text: string): boolean {
  const opens = { '(': ')', '[': ']', '{': '}' };
  const stack: string[] = [];
  for (const ch of text) {
    if (ch in opens) stack.push(opens[ch as keyof typeof opens]);
    if ([')', ']', '}'].includes(ch)) {
      if (stack.length > 0 && stack[stack.length - 1] === ch) stack.pop();
    }
  }
  return stack.length > 0;
}

function hasUnclosedQuotes(text: string): boolean {
  let single = 0;
  let double = 0;
  for (const ch of text) {
    if (ch === "'") single++;
    if (ch === '"') double++;
  }
  return single % 2 !== 0 || double % 2 !== 0;
}

function hasIncompleteCodeBlock(text: string): boolean {
  const opens = (text.match(/```/g) || []).length;
  return opens % 2 !== 0;
}

function endsWithEllipsis(text: string): boolean {
  const trimmed = text.trimEnd();
  return trimmed.endsWith('...') || trimmed.endsWith('\u2026');
}

function hasIncompleteList(text: string): boolean {
  const lines = text.trimEnd().split('\n');
  const last = lines[lines.length - 1]?.trim() ?? '';
  return /^\d+\.\s*$/.test(last) || /^[-*]\s*$/.test(last);
}

export function outputTruncation(options: OutputTruncationOptions): Guard {
  return {
    name: 'output-truncation',
    version: '0.1.0',
    description: 'Detects if output was likely truncated or incomplete',
    category: 'format',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const signals: string[] = [];

      if (endsMidSentence(text)) signals.push('ends-mid-sentence');
      if (hasUnclosedBrackets(text)) signals.push('unclosed-brackets');
      if (hasUnclosedQuotes(text)) signals.push('unclosed-quotes');
      if (hasIncompleteCodeBlock(text)) signals.push('incomplete-code-block');
      if (endsWithEllipsis(text)) signals.push('ends-with-ellipsis');
      if (hasIncompleteList(text)) signals.push('incomplete-list');

      const triggered = signals.length > 0;
      const score = triggered ? Math.min(signals.length / 3, 1.0) : 0;

      return {
        guardName: 'output-truncation',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { signals } : undefined,
      };
    },
  };
}
